planner/optimize.go

```diff
@@ -118,109 +119,132 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 			return fp, fp.OutputNames(), nil
 		}
 	}
-
 	sctx.PrepareTSFuture(ctx)
 
-	bestPlan, names, _, err := optimize(ctx, sctx, node, is)
-	if err != nil {
-		return nil, nil, err
-	}
-	if !(sessVars.UsePlanBaselines || sessVars.EvolvePlanBaselines) {
-		return bestPlan, names, nil
-	}
+	useBinding := sessVars.UsePlanBaselines
 	stmtNode, ok := node.(ast.StmtNode)
 	if !ok {
-		return bestPlan, names, nil
-	}
-	bindRecord, scope, err := getBindRecord(sctx, stmtNode)
-	if err != nil {
-		return nil, nil, err
-	}
-	if bindRecord == nil {
-		return bestPlan, names, nil
-	}
-	if sctx.GetSessionVars().SelectLimit != math.MaxUint64 {
-		sctx.GetSessionVars().StmtCtx.AppendWarning(errors.New("sql_select_limit is set, so plan binding is not activated"))
-		return bestPlan, names, nil
-	}
-	err = setFoundInBinding(sctx, true)
-	if err != nil {
-		return nil, nil, err
-	}
-	bestPlanHint := plannercore.GenHintsFromPhysicalPlan(bestPlan)
-	if len(bindRecord.Bindings) > 0 {
-		orgBinding := bindRecord.Bindings[0] // the first is the original binding
-		for _, tbHint := range tableHints {  // consider table hints which contained by the original binding
-			if orgBinding.Hint.ContainTableHint(tbHint.HintName.String()) {
-				bestPlanHint = append(bestPlanHint, tbHint)
-			}
+		useBinding = false
+	}
+	if useBinding && sessVars.SelectLimit != math.MaxUint64 {
+		sessVars.StmtCtx.AppendWarning(errors.New("sql_select_limit is set, ignore SQL bindings"))
+		useBinding = false
+	}
+	var (
+		bindRecord *bindinfo.BindRecord
+		scope      string
+		err        error
+	)
+	if useBinding {
+		bindRecord, scope, err = getBindRecord(sctx, stmtNode)
+		if err != nil || bindRecord == nil || len(bindRecord.Bindings) == 0 {
+			useBinding = false
 		}
 	}
-	bestPlanHintStr := hint.RestoreOptimizerHints(bestPlanHint)
 
-	defer func() {
-		sessVars.StmtCtx.StmtHints = stmtHints
-		for _, warn := range warns {
-			sctx.GetSessionVars().StmtCtx.AppendWarning(warn)
-		}
-	}()
-	binding := bindRecord.FindBinding(bestPlanHintStr)
-	// If the best bestPlan is in baselines, just use it.
-	if binding != nil && binding.Status == bindinfo.Using {
-		if sctx.GetSessionVars().UsePlanBaselines {
-			stmtHints, warns = handleStmtHints(binding.Hint.GetFirstTableHints())
+	var names types.NameSlice
+	var bestPlan, bestPlanFromBind plannercore.Plan
+	if useBinding {
+		minCost := math.MaxFloat64
+		var bindStmtHints stmtctx.StmtHints
+		originHints := hint.CollectHint(stmtNode)
+		// bindRecord must be not nil when coming here, try to find the best binding.
+		for _, binding := range bindRecord.Bindings {
+			if binding.Status != bindinfo.Using {
+				continue
+			}
+			metrics.BindUsageCounter.WithLabelValues(scope).Inc()
+			hint.BindHint(stmtNode, binding.Hint)
+			curStmtHints, _, curWarns := handleStmtHints(binding.Hint.GetFirstTableHints())
+			sessVars.StmtCtx.StmtHints = curStmtHints
+			plan, curNames, cost, err := optimize(ctx, sctx, node, is)
```

Comment:
I have a question here: if there are many bindings for an origin SQL, it still needs to optimize many times?

Reply:
Yes, currently baseline evolution is not GA, and we have not refactored the create binding behavior, so one original_sql would only have one binding.